crates/shielded_token/src/conversion.rs

```diff
@@ -226,18 +245,236 @@ where
+#[allow(clippy::too_many_arguments)]
+fn update_non_native_conversions<S, TransToken>(
+    storage: &mut S,
+    token: &Address,
+    ref_inflation: u128,
+    normed_inflation: u128,
+    masp_epochs_per_year: u64,
+    masp_epoch: MaspEpoch,
+    reward_assets: [AssetType; 4],
+    total_reward: &mut Amount,
+    current_convs: &mut BTreeMap<
+        (Address, Denomination, MaspDigitPos),
+        AllowedConversion,
+    >,
+) -> Result<Denomination>
+where
+    S: StorageWrite + StorageRead + WithConversionState,
+    TransToken:
+        trans_token::Keys + trans_token::Read<S> + trans_token::Write<S>,
+{
+    let prev_masp_epoch =
+        masp_epoch.prev().ok_or_err_msg("MASP epoch underflow")?;
+    let (precision, denom) =
+        calculate_masp_rewards_precision::<S, TransToken>(storage, token)?;
+    let (reward, precision) = calculate_masp_rewards::<S, TransToken>(
+        storage,
+        token,
+        denom,
+        precision,
+        masp_epochs_per_year,
+    )?;
+    // Express the inflation reward in real terms, that is, with
+    // respect to the native asset in the zeroth epoch
+    let reward_uint = Uint::from(reward);
+    let ref_inflation_uint = Uint::from(ref_inflation);
+    let inflation_uint = Uint::from(normed_inflation);
+    let real_reward =
+        checked!((reward_uint * ref_inflation_uint) / inflation_uint)?
+            .try_into()
+            .unwrap_or_else(|_| {
+                tracing::warn!(
+                    "MASP reward for {} assumed to be 0 because the computed \
+                     value is too large. Please check the inflation \
+                     parameters.",
+                    token
+                );
+                0u128
+            });
+    // The conversion is computed such that if consecutive
+    // conversions are added together, the
+    // intermediate tokens cancel/ telescope out
+    let precision_i128 = i128::try_from(precision).into_storage_result()?;
+    let real_reward_i128 = i128::try_from(real_reward).into_storage_result()?;
+    for digit in MaspDigitPos::iter() {
+        // Provide an allowed conversion from previous timestamp. The
+        // negative sign allows each instance of the old asset to be
+        // cancelled out/replaced with the new asset
+        let old_asset = encode_asset_type(
+            token.clone(),
+            denom,
+            digit,
+            Some(prev_masp_epoch),
+        )
+        .into_storage_result()?;
+        let new_asset =
+            encode_asset_type(token.clone(), denom, digit, Some(masp_epoch))
+                .into_storage_result()?;
+
+        current_convs.insert(
+            (token.clone(), denom, digit),
+            checked!(
+                MaspAmount::from_pair(old_asset, -precision_i128)
+                    + &MaspAmount::from_pair(new_asset, precision_i128)
+                    + &MaspAmount::from_pair(
+                        reward_assets[digit as usize],
```

Comment:
This assumes that a reward for, e.g., digit 2 will automatically be given out as an amount of digit 2 native token, is that right?

Reply:
Yes, exactly. Each allowed conversion put into the Merkle tree deals with exactly one digit. No mixing of digits is currently happening.